### ISSUE TITLE
Question about the ablation study

### ISSUE THREAD

**Issue by biubiubiiu** [swz30/Restormer#37]
Thanks for your outstanding work. I have some confusion about the implementation of group(b) and group(e) in the ablation study in Sec. 4.5. For group(b), with MTA+standard FN, should I break the feature map into patches before passing it to the fully-connected layer and reshape it to its original size after FC. If so, what's the patch size used, and is dropout applied here? And for group(e), what is the structure of DFN since the gating mechanism is removed? Is it just a sequential LayerNorm and 1*1 convolution?

**Comment by swz30** [comment]
Hi @biubiubiiu 

Here are the implementations:

**Multi-Head Transposed Self-Attention (MTA)**
```
class Attention(nn.Module):	
    def __init__(self, dim, num_heads, bias):	
        super(Attention, self).__init__()	
        self.num_heads = num_heads	
        self.temperature = nn.Parameter(torch.ones(num_heads, 1, 1))	
        self.qkv = nn.Conv2d(dim, dim*3, kernel_size=1, bias=bias)	
        self.project_out = nn.Conv2d(dim, dim, kernel_size=1, bias=bias)	
        	
    def forward(self, x):	
        b,c,h,w = x.shape	
        qkv = self.qkv(x)	
        q,k,v = qkv.chunk(3, dim=1)   	
        	
        q = rearrange(q, 'b (head c) h w -> b head c (h w)', head=self.num_heads)	
        k = rearrange(k, 'b (head c) h w -> b head c (h w)', head=self.num_heads)	
        v = rearrange(v, 'b (head c) h w -> b head c (h w)', head=self.num_heads)	
        q = torch.nn.functional.normalize(q, dim=-1)	
        k = torch.nn.functional.normalize(k, dim=-1)	
        attn = (q @ k.transpose(-2, -1)) * self.temperature	
        attn = attn.softmax(dim=-1)	
        out = (attn @ v)	
        	
        out = rearrange(out, 'b head c (h w) -> b (head c) h w', head=self.num_heads, h=h, w=w)	
        out = self.project_out(out)	
        return out
```

**Feed-forward Network (FN) in the group (b)**	
```
class FeedForward(nn.Module):	
    def __init__(self, dim, ffn_expansion_factor, bias):	
        super(FeedForward, self).__init__()	
        hidden_features = int(dim*ffn_expansion_factor)	
        self.project_in = nn.Conv2d(dim, hidden_features, kernel_size=1, bias=bias)		
        self.project_out = nn.Conv2d(hidden_features, dim, kernel_size=1, bias=bias)	
    def forward(self, x):	
        x = self.project_in(x)		
        x = F.gelu(x)	
        x = self.project_out(x)	
        return x	
```

**Dconv Feed-forward Network (DFN) in the group (e)**
```
## Feed-forward Network	
class FeedForward(nn.Module):	
    def __init__(self, dim, ffn_expansion_factor, bias):	
        super(FeedForward, self).__init__()	
        hidden_features = int(dim*ffn_expansion_factor)	
        self.project_in = nn.Conv2d(dim, hidden_features, kernel_size=1, bias=bias)	
        self.dwconv = nn.Conv2d(hidden_features, hidden_features, kernel_size=3, stride=1, padding=1, groups=hidden_features, bias=bias)	
        self.project_out = nn.Conv2d(hidden_features, dim, kernel_size=1, bias=bias)	
    def forward(self, x):	
        x = self.project_in(x)	
        x = self.dwconv(x)	
        x = F.gelu(x)	
        x = self.project_out(x)	
        return x	
```

For both ablations (b) and (e), we use ``` ffn_expansion_factor = 4``` .
